### DISCUSSION TITLE
hi👋, can i run App in background on MacOS? i want close main window but not kill process.

### DISCUSSION THREAD

**Discussion by ChrisChou-freeman** [beeware/toga#1437] [Q&A] [answered]
i want develop an app on MacOS, with Statusbar function, which i use [rumps](https://github.com/jaredks/rumps), but every time i close MainWindow, the whole program exits. can i stop toga App quit program just run in background.

Thanks🙏

**Comment by freakboy3742** [comment, marked as answer]
Right now, there isn't a straightforward way to do this. 

You might be able to hack something together by subclassing toga.App and overwriting some of the base methods in creative ways to disable MainWindow creation. This would create an app, but without a window that closes. It might also be worthwhile looking at toga.DocumentApp as a starting point for this hacking, as DocumentApp doesn't have a MainWindow by default.

Longer term, "statusbar widget" apps are a feature we have long wanted to support (See #97); if someone were to propose an API and implementation (even if that implementation only existed on one platform), we'd consider merging it.

**Reply by ChrisChou-freeman** [reply]
thanks, i will try it.

**Reply by Rybec** [reply]
I have a very closely related question:

I want to use Beeware to write a locally running web server based on Python's http.server, but obviously it needs to run in the background, so that a browser can access it in the foreground.  It doesn't need any kind of window or rendering.

Is the answer the same for this, or is there some simpler way to do this?


On a side note: For anyone who might be looking at adding this feature, assuming it doesn't already exist in some way, Termux _can be used_ to run Python HTTP servers in the background (but this isn't a feasible option for regular people).  Maybe look at the Termux code to figure out how to do this: [Termux on Github](https://github.com/termux/termux-app).

**Reply by freakboy3742** [reply]
@Rybec Yes, the answer is essentially the same. Briefcase is optimised for building desktop apps; desktop apps have windows (or, at least, a manifestation on the toolbar). You may be be able to subclass toga.App to remove the parts that expect to create a MainWindow at startup, and use that main window as a trigger for shutting down the app.

**Reply by Rybec** [reply]
Alright, thanks!  Looks like python-on-android might work better for my uses right now then, but I'll keep an eye on this!

**Reply by freakboy3742** [reply]
Wait - what? The original question was about macOS. If you're talking about building an Android app, then the constraints are *completely* different. User-space phone apps *require* a UI, and the concept of a "background task" is radically different. It might be possible to set up a web server in a separate process, but it's going to be a very complex (and probably unreliable) thing to set up.

**Reply by Rybec** [reply]
Lol!  I clearly misunderstood the question.  I assumed _development_ was happening on MacOS, but rereading the title, it's clear I was wrong!

Never mind then.  Sorry I didn't read the title very well before asking.

**Reply by mhsmith** [reply]
I don't know about the background task limitations on iOS, but this shouldn't be too hard on Android. If you're using the basic `http.server` API, then you could run it in a background thread of the main app process: there's no need to create a separate process.

The tricky part is what happens when you switch to the browser and the server process goes into the background. By default the server will be treated as a "cached process" (level 4 on [this list](https://developer.android.com/guide/components/activities/process-lifecycle)). Android won't shut it down immediately, but it can be killed any time the system needs to free up memory. Depending on your use case, this might be good enough.

If you need to give the server more priority, then you can modify the Android project to create a Service, or a foreground Service, which will elevate it to level 3 or 2 respectively on [this list](https://developer.android.com/guide/components/activities/process-lifecycle). The service doesn't have to actually do anything, it just has to be started. BeeWare doesn't have any direct support for this, but you can find lots of example code on the Android developer website and elsewhere.

**Reply by Rybec** [reply]
Thanks for the information!  I'm trying to avoid using Java entirely.  I've done some Android development in Java, and honestly, the stuff I did in C was easier, and that includes over a month working out how to setup a good development environment for that.  I know Kivy and python-on-android have some access to Android APIs, so it's possible they have this.  And of course, maybe I could find a way to make this work in BeeWare as well.  This at least gives me a solid starting point though, so thank you again!

**Reply by freakboy3742** [reply]
FWIW - iOS background task restrictions are quite different; they are mostly limited to APIs where you can register that your app has an interest in external information (e.g., I want to be notified when the geolocation moves into a given region), at which point a callback can be invoked. AFAIK, there's no analog of a "run this web server loop in the background" type service.

I'll also add that my experience with Android background services is that they're also a great way to drain your battery - at least, it is if you take the naive approach of "just run a loop in the background" :-) Anything that is running an active loop in the background means your phone doesn't ever truly go to sleep. Yes, it works, and with service level elevation you can stop the OS from killing the worker; but your app's performance won't be that great unless you make some optimisations around how your service works so that the phone can be *not* working unless something is actually happening.

**Reply by Rybec** [reply]
Python's http.server puts an OS managed listener on the port, rather than polling.  It's incredibly light weight, and it's using event based blocking rather than a polling loop.  And I've actually used it on Android before, in Termux, which _will_ run it in the background, without any battery issues.  If this was just for personal use (like my last local Android web app), I would just use Python in Termux again, but this project is one that could end up having commercial value.  The last thing I need is a system where the user has to have significant technical skills to use the product.

It looks like python-for-android does have support for services, so I'll probably use that for now.  It looks like BeeWare is more suited to front end apps rather than services, so if I ever get around to developing one of the many other apps I want to write, I may be back!  (Actually, I'm writing a desktop game in Python that I might want to port to mobile...  I don't think that will happen very soon though.)
